pkg/watchdog/watchdog.go

```diff
@@ -0,0 +1,228 @@
+)
+
+// Option are options passed to SetOptions().
+type Option int32
+
+// Bitset of possible flags for the Option type.
+const (
+	// Unknown status error
+	OptionUnknown Option = -1
+	// Turn off the watchdog timer
+	OptionDisableCard Option = 0x0001
+	// Turn on the watchdog timer
+	OptionEnableCard Option = 0x0002
+	// Kernel panic on temperature trip
+	OptionTempPanic Option = 0x0004
+)
+
+type Watchdog struct {
+	f *os.File
+}
+
+func Open(dev string) (*Watchdog, error) {
+	f, err := os.OpenFile(dev, os.O_RDWR, 0)
+	if err != nil {
+		return nil, err
+	}
+	return &Watchdog{f: f}, nil
+}
+
+// Close closes the device without disarming the watchdog.
+func (w *Watchdog) Close() error {
+	return w.f.Close()
+}
+
+// MagicClose disarms the watchdog. However if the kernel is compiled with
+// CONFIG_WATCHDOG_NOWAYOUT=y, there may be no way to disarm the watchdog.
+func (w *Watchdog) MagicClose() error {
+	if _, err := w.f.Write([]byte("V")); err != nil {
+		w.f.Close()
+		return err
+	}
+	return w.f.Close()
+}
+
+// Support returns the WatchdogInfo struct.
+func (w *Watchdog) Support() (*unix.WatchdogInfo, error) {
+	var wi unix.WatchdogInfo
+	if _, _, err := unix.Syscall(unix.SYS_IOCTL, w.f.Fd(), wdiocGetSupport, uintptr(unsafe.Pointer(&wi))); err != 0 {
+		return nil, err
+	}
+	return &wi, nil
+}
+
+// Status returns the current status.
+func (w *Watchdog) Status() (Status, error) {
+	flags, err := unix.IoctlGetUint32(int(w.f.Fd()), wdiocGetStatus)
+	if err != nil {
+		return StatusUnknown, err
+	}
+	return Status(flags), nil
+}
+
+// BootStatus returns the status at the last reboot.
+func (w *Watchdog) BootStatus() (Status, error) {
+	flags, err := unix.IoctlGetUint32(int(w.f.Fd()), wdiocGetBootStatus)
+	if err != nil {
+		return StatusUnknown, err
+	}
+	return Status(flags), nil
+}
+
+// Temp returns temperature in degrees Fahrenheit.
+func (w *Watchdog) Temp() (int32, error) {
+	temp, err := unix.IoctlGetUint32(int(w.f.Fd()), wdiocGetTemp)
+	if err != nil {
+		return 0, err
+	}
+	return int32(temp), nil
+}
+
+// SetOptions can be used to control some aspects of the cards operation.
+func (w *Watchdog) SetOptions(options Option) error {
+	if _, _, err := unix.Syscall(unix.SYS_IOCTL, w.f.Fd(), wdiocSetOptions, uintptr(unsafe.Pointer(&options))); err != 0 {
+		return err
+	}
+	return nil
+}
+
+// KeepAlive pets the watchdog.
+func (w *Watchdog) KeepAlive() error {
+	_, err := w.f.WriteString("1")
+	return err
+}
+
+// SetTimeout sets the watchdog timeout on the fly. It returns the real timeout
+// used and this timeout might differe from the requested one due to limitation
+// of the hardware.
+func (w *Watchdog) SetTimeout(timeout int32) (int32, error) {
+	originalTimeout := timeout
+	if _, _, err := unix.Syscall(unix.SYS_IOCTL, w.f.Fd(), wdiocSetTimeout, uintptr(unsafe.Pointer(&timeout))); err != 0 {
+		return 0, err
+	}
+	if originalTimeout != timeout {
+		log.Printf("Watchdog timeout set to %ds, wanted %ds", timeout, originalTimeout)
```

Comment:
a library printing logs is usually not nice

return an error message? user can decide to ignore it.